signature/simplesigning/signer_test.go

```diff
@@ -0,0 +1,234 @@
+	// should manifest even with a valid passphrase or unlocked key (because the GPG agent is caching unlocked keys).
+	// Alternatively, we could be caling gpgagent.KillGPGAgent() all the time...
+	type failingCase struct {
+		name string
+		opts []Option
+		// NOTE: We DO NOT promise that things that don't fail during NewSigner won't start failing there.
+		// Actually we’d prefer failures to be identified early. This field only records current expected behavior, not the _desired_ end state.
+		creationFails         bool
+		creationErrorContains string
+		manifest              []byte
+		ref                   reference.Named
+	}
+	testFailure := func(c failingCase) {
+		s, err := NewSigner(c.opts...)
+		if c.creationFails {
+			assert.Error(t, err, c.name)
+			if c.creationErrorContains != "" {
+				assert.ErrorContains(t, err, c.creationErrorContains, c.name)
+			}
+		} else {
+			require.NoError(t, err, c.name)
+			defer s.Close()
+
+			m := manifest
+			if c.manifest != nil {
+				m = c.manifest
+			}
+			_, err = internalSigner.SignImageManifest(s, m, c.ref)
+			assert.Error(t, err, c.name)
+		}
+	}
+	for _, c := range []failingCase{
+		{
+			name: "Invalid passphrase",
+			opts: []Option{
+				WithKeyFingerprint(testKeyFingerprintWithPassphrase),
+				WithPassphrase(testPassphrase + "\n"),
+			},
+			creationFails:         true,
+			creationErrorContains: "invalid passphrase",
+			ref:                   testImageSignatureReference,
+		},
+		{
+			name: "Wrong passphrase",
+			opts: []Option{
+				WithKeyFingerprint(testKeyFingerprintWithPassphrase),
+				WithPassphrase("wrong"),
+			},
+			ref: testImageSignatureReference,
+		},
+		{
+			name: "No passphrase",
+			opts: []Option{WithKeyFingerprint(testKeyFingerprintWithPassphrase)},
+			ref:  testImageSignatureReference,
+		},
+	} {
+		testFailure(c)
+	}
+
+	// Successful signing
+	for _, c := range []struct {
+		name        string
+		fingerprint string
+		opts        []Option
+	}{
+		{
+			name:        "No passphrase",
+			fingerprint: testKeyFingerprint,
+		},
+		{
+			name:        "With passphrase",
+			fingerprint: testKeyFingerprintWithPassphrase,
+			opts:        []Option{WithPassphrase(testPassphrase)},
+		},
+	} {
+		s, err := NewSigner(append([]Option{WithKeyFingerprint(c.fingerprint)}, c.opts...)...)
+		require.NoError(t, err, c.name)
+		defer s.Close()
+
+		sig, err := internalSigner.SignImageManifest(s, manifest, testImageSignatureReference)
+		require.NoError(t, err, c.name)
+		simpleSig, ok := sig.(internalSig.SimpleSigning)
+		require.True(t, ok)
+
+		// FIXME FIXME: gpgme_op_sign with a passphrase succeeds, but somehow confuses the GPGME internal state
+		// so that gpgme_op_verify below never completes (it polls on an already closed FD).
+		// That’s probably a GPG bug, and needs investigating and fixing, but it isn’t related to this “signer” implementation.
+		if len(c.opts) == 0 {
```

Comment:
This is a bug that certainly needs tracking down: https://github.com/containers/image/issues/1777 .

“Luckily” it probably only affects long-term processes that sign (and verify?), i.e. not Podman and not CRI-O.